Add setting "system" category forms and tests

marshmalien opened the pull request:
### SUMMARY
Activity Stream: [https://github.com/ansible/awx/issues/8287](https://github.com/ansible/awx/issues/8287)
Logging: [https://github.com/ansible/awx/issues/8292](https://github.com/ansible/awx/issues/8292)
Misc System: [https://github.com/ansible/awx/issues/8293](https://github.com/ansible/awx/issues/8293)
AzureAD: https://github.com/ansible/awx/issues/8286

Adds "System" setting category forms: Activity Stream, Logging, and Miscellaneous System. The patterns used in this area are repeated in all setting forms, so I would like feedback on the technical approach and UX. 

The new field revert button will behave differently from the old ui revert button. In the old UI, clicking the field-level revert button executes the request immediately; it sends a PATCH request to the API without requiring the user to confirm the action in an alert modal. In the new UI, the revert button does not execute the request immediately; instead, the button will update the form with the system default value and switch the button text from "revert" to "undo". If the user is happy with their selection, they can choose to save the form. If the user selected revert by accident, the undo button will allow the user to recover from their action by reverting the field value back to the initial form value. The revert button will appear disabled if the current, initial, and default field values are all equal.

![setting-undo](https://user-images.githubusercontent.com/15881645/97365122-8cf76680-187b-11eb-9f48-521bd325ec37.gif)

### RBAC
To guarantee that only superusers have permission to view setting forms, the edit route checks that the user has system administration privileges before rendering the edit component. If the user does not have system administration privileges, a Redirect will take the user back to the setting details view.

```jsx
{me?.is_superuser ? (
  <ActivityStreamEdit />
) : (
  <Redirect to={`${baseURL}/details`} />
)}
```

### Additional components and hook
- useModal - handles opening and closing a modal
- RevertAllAlert modal - Confirmation alert modal shown when the user chooses to "Revert all".
- RevertButton - Button that handles reverting the form field value to either the default system setting value or initial form value.
- RevertFormActionGroup - Form action buttons that include Save, Revert all to default, and Cancel by default.
- LoggingTestAlert toast - Toast alert with the status of the test log message.

![test-log](https://user-images.githubusercontent.com/15881645/97365257-c7610380-187b-11eb-869e-f00a280571eb.gif)
![test-log-err](https://user-images.githubusercontent.com/15881645/97365267-ca5bf400-187b-11eb-8462-06754b58d1e7.gif)

marshmalien committed:
Add setting system category forms and tests

 * Add activity stream, logging, and misc system forms
 * Hookup logging test alert
 * Hookup revert buttons
 * Add useModal helper hook
 * Swap VariablesDetail for CodeDetail within setting detail views
 * Update SettingDetail import path in setting detail views

marshmalien commented on awx/ui_next/src/components/FormActionGroup/FormActionGroup.jsx:
```diff
@@ -24,16 +18,6 @@ const FormActionGroup = ({
         >
           {i18n._(t`Save`)}
         </Button>
-        {onRevert && (
-          <Button
-            aria-label={i18n._(t`Revert`)}
-            variant="secondary"
-            type="button"
-            onClick={onRevert}
-          >
-            {i18n._(t`Revert`)}
-          </Button>
-        )}
```
Since the ability to revert is a settings-only feature, I removed the onRevert button from the top-level FormActionGroup component and created a custom  action button group for Settings called RevertFormActionGroup. The RevertFormActionGroup component allows the form to pass in additional action buttons as `{ children }` and keeps the top-level FormActionGroup generic.

jlmitch5 commented on awx/ui_next/src/api/models/Settings.js:
```diff
@@ -21,6 +21,10 @@ class Settings extends Base {
     return this.http.options(`${this.baseUrl}${category}/`);
   }
+
+  test(category, data) {
```
I wonder if we should attempt to rename this with a verb so the REST action is clear...createTest maybe?

AlexSCorey replied:
I agree, I think this should be createTest as well.

jlmitch5 commented on awx/ui_next/src/screens/Setting/shared/RevertAllAlert.jsx:
```diff
@@ -0,0 +1,39 @@
+        </Button>,
+      ]}
+    >
+      {i18n._(t`This will revert all configuration values to their
```
Should we indicate in this message that revert all is for the particular form you are on, not every setting configuration value across all forms?

marshmalien replied:
I consulted with @tvo318 on this one. What do you think about:
```
This will revert all configuration values on this page to their factory defaults. Are you sure you want to proceed?
```

jlmitch5 replied:
Yes that's good!

jlmitch5 commented on awx/ui_next/src/screens/Setting/shared/RevertButton.jsx:
```diff
@@ -0,0 +1,66 @@
+
+const ButtonWrapper = styled.div`
+  margin-left: auto;
+  &&& {
```
TIL about this trick for bumping the specificity, cool stuff

jlmitch5 commented on awx/ui_next/src/screens/Setting/shared/SharedFields.jsx:
```diff
@@ -0,0 +1,258 @@
+  isRequired: bool,
+};
+
+export { BooleanField, ChoiceField, EncryptedField, InputField, ObjectField };
```
very cool we've been able to keep this pseudo-form generator very tight!

jlmitch5 commented on awx/ui_next/src/screens/Setting/shared/ShareFields.test.jsx:
```diff
@@ -0,0 +1,152 @@
+import React from 'react';
```
This file is supposed to be named "SharedFields" I think

marshmalien replied:
😄 I think you're right.

jlmitch5 commented on awx/ui_next/src/screens/Setting/ActivityStream/ActivityStreamEdit/ActivityStreamEdit.jsx:
```diff
@@ -1,25 +1,137 @@
+
+function ActivityStreamEdit() {
+  const history = useHistory();
+  const { isModalOpen, toggleModal, closeModal } = useModal();
```
Slick way to handle this kind of thing!

jlmitch5 commented on awx/ui_next/src/screens/Setting/Logging/LoggingEdit/LoggingEdit.test.jsx:
```diff
@@ -1,16 +1,318 @@
+    await waitForElement(wrapper, 'ContentLoading', el => el.length === 0);
+    expect(wrapper.find('ContentError').length).toBe(1);
+  });
 });
```
extremely thorough tests here!

jlmitch5 reviewed (commented):
Looks great!  I think the settings code has turned out incredibly simple for how much is going on, great job.  Just a few small things, mostly UX/wording related and not really code related.

AlexSCorey commented on awx/ui_next/src/screens/Setting/AzureAD/AzureADEdit/AzureADEdit.jsx:
```diff
@@ -1,25 +1,146 @@
+
+  useEffect(() => {
+    if (submitResult) {
+      history.push('/settings/azure/details');
```
Curious why you are navigating this way instead of doing the navigation inside of the useRequest [here](https://github.com/ansible/awx/pull/8472/files#diff-eb84bb28b0549e1934e10a2c5de60ac40f5224fccd28a4a414fa0813c65b0c8bR50).  Is it possible that a submitResult not be returned?

marshmalien replied:
Both patterns achieve the same goal. It is possible that a submitResult not be returned if there is an error or if the request function does not return a result. I've seen both patterns used in our codebase, but I'm open to moving navigation into the request function, for brevity.

AlexSCorey commented on awx/ui_next/src/screens/Setting/ActivityStream/ActivityStreamEdit/ActivityStreamEdit.jsx:
```diff
@@ -1,25 +1,137 @@
+    error,
+    request,
+    result: {
+      ACTIVITY_STREAM_ENABLED,
```
I'm seeing this all caps, constant syntax more and more.  Is this the direction the team is moving in?  I like it and think it makes sense.

marshmalien replied:
This is all caps because the fields are all caps in the API response. I wasn't trying to change the syntax, more so, I was trying to avoid doing a destructuring reassignment for the sake of casing.

nixocio commented on awx/ui_next/src/screens/Setting/Logging/LoggingEdit/LoggingEdit.jsx:
```diff
@@ -1,23 +1,272 @@
+                      name="LOG_AGGREGATOR_TCP_TIMEOUT"
+                      config={logging.LOG_AGGREGATOR_TCP_TIMEOUT}
+                      type="number"
+                      isRequired
```
What about we add more specific validator for `type="number"` ? We do have `number` or `integer` validators.

![image](https://user-images.githubusercontent.com/9053044/97589995-8e887200-19d4-11eb-8f5a-3e189292df35.png)

marshmalien replied:
Of course Firefox will let us type non-numeric characters in a `type="number"` input field 🙃See:https://bugzilla.mozilla.org/show_bug.cgi?id=1398528
If I remove the type attribute from TextInput, the validators will work as expected.
![input](https://user-images.githubusercontent.com/15881645/97628257-d625f280-1a02-11eb-8a74-52107c7aff10.gif)

AlexSCorey reviewed (commented):
I think this looks really good. I can the tests locally and they all pass. You were able to address all of my concerns offline.  I did find some thing about the tooltips, which you said you'd look into.  Overall, awesome work!

nixocio reviewed (commented):
Great job @marshmalien. I got this code locally, and it seems to be working as expected. I left a minor suggestion.

marshmalien committed:
Fix number input validation bug

unlikelyzero commented:
Feedback:
- If entering an incorrect format into a CodeMirror component (i.e. Azure AD Map), the except is thrown to console without feedback to the user
- There doesn't seem to be a unique attribute associated with the individual 'Reverts' or 'Undo'. Is that something you could add via ID or aria?
- Error on when clicking into `/next/settings/miscellaneous_system/edit`page. 
```
TypeError: Cannot set property 'value' of undefined at https://localhost:3001/static/js/main.chunk.js:87121:29 at Array.forEach (<anonymous>) at https://localhost:3001/static/js/main.chunk.js:87119:29 at async https://localhost:3001/static/js/main.chunk.js:113038:26
```

marshmalien committed:
Do not render setting field if config is empty

jlmitch5 reviewed (approved):
looks good!

marshmalien commented:
@unlikelyzero 
> If entering an incorrect format into a CodeMirror component (i.e. Azure AD Map), the except is thrown to console without feedback to the user.

Updated PR to correctly handle code field error messages.
<img width="457" alt="Screen Shot 2020-11-03 at 2 36 17 PM" src="https://user-images.githubusercontent.com/15881645/98032451-08a86480-1de2-11eb-8210-33fb3054cbae.png">

>There doesn't seem to be a unique attribute associated with the individual 'Reverts' or 'Undo'. Is that something you could add via ID or aria?

All revert buttons should have an aria-label="Revert". I added a data-cy prop that concats the field id and "-revert" together. Example: `data-cy=AUTOMATION_ANALYTICS_URL-revert`. 

>Error on when clicking into /next/settings/miscellaneous_system/edit page.

Updated PR to handle rendering empty fields.

marshmalien committed:
Use ouiaId instead of data-cy id in RevertButton

quarckster commented on awx/ui_next/src/screens/Setting/shared/RevertButton.jsx:
```diff
@@ -45,7 +45,7 @@ function RevertButton({ i18n, id, defaultValue, isDisabled = false }) {
         <Button
           aria-label={isRevertable ? i18n._(t`Revert`) : i18n._(t`Undo`)}
-          data-cy={`${id}-revert`}
+          data-ouia-id={`${id}-revert`}
```
it should be `ouiaId`:
```js
ouiaId={`${id}-revert`}
```
during the app building it will be transformed into `data-ouia-component-id`

marshmalien replied:
Thanks! I can confirm that adding `ouiaId` to the Revert Button transforms into the following additional attributes:
* data-ouia-component-type="PF4/Button" 
* data-ouia-safe="true" 
* data-ouia-component-id="TOWER_URL_BASE-revert"

unlikelyzero commented:
Need to update this line to provide better instruction to user on how the main logging toggle can be enabled. https://github.com/ansible/awx/blob/10664d1931d789c281e2c9d3d2f14c0a69cb9e13/awx/main/conf.py#L659

marshmalien committed:
Add descriptive helper text when log aggregator toggle is disabled

unlikelyzero commented:
recheck

unlikelyzero commented:
@marshmalien just a single defect was found on this last pass:
'Reverting' the 'Base URL of the Tower host' under /Miscellaneous_system/edit changes the URL to localhost

marshmalien commented:
@unlikelyzero Is that different from the old UI? The field should revert to the same value that is returned from `/api/v2/settings/all/` OPTIONS response. 
```
"TOWER_URL_BASE": {
  "type": "string",
  "required": true,
  "label": "Base URL of the Tower host",
  "help_text": "This setting is used by services like notifications to render a valid url to the Tower host.",
  "category": "System",
  "category_slug": "system",
  "default": "https://localhost:3001"
},
```

unlikelyzero commented:
covered by https://github.com/ansible/tower-qa/pull/5734
